File: 0101-0200/0103.Binary_Tree_Zigzag_Level_Order_Traversal.js
//0ms

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function (root) {
    if (!root) return [];
    const que = [root];
    const res = [];
    let flag = true;

    while (true) {
        if (que.length === 0) break;
        const currLevel = [];
        const len = que.length;
        for (let i = 0; i < len; i++) {
            let temp = que.shift();
            if (temp) {
                currLevel.push(temp.val);
                if (temp.left) que.push(temp.left);
                if (temp.right) que.push(temp.right);
            }
        }
        if (flag) {
            res.push(currLevel);
        } else {
            res.push(currLevel.reverse());
        }
        flag = !flag;
    }
    return res;
};

File: 0101-0200/0103.Binary_Tree_Zigzag_Level_Order_Traversal.java
//0ms


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 
 public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        oie(root, res,0);
        return res;
    }

    public void oie(TreeNode node, List<List<Integer>> res, int level)
    {
        if(node==null)   return;

        if(res.size()<=level)
        {
            res.add(new LinkedList<>());
        }
         List<Integer> commando = res.get(level);

         if(level%2 == 0)
         {
            commando.add(node.val);
         }
         else{
            commando.add(0,node.val);
         }

         oie(node.left,res,level+1);
         oie(node.right, res, level+1);
    }
 }














//  public class Solution {
//     public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
//         List<List<Integer>> sol = new ArrayList<>();
//         travel(root, sol, 0);
//         return sol;
//     }

//     private void travel(TreeNode curr, List<List<Integer>> sol, int level) {
//         if (curr == null) return;
        
//         if (sol.size() <= level) {
//             sol.add(new LinkedList<>());
//         }
        
//         List<Integer> collection = sol.get(level);
        
//         if (level % 2 == 0) {
//             collection.add(curr.val);
//         } else {
//             collection.add(0, curr.val);
//         }

//         travel(curr.left, sol, level + 1);
//         travel(curr.right, sol, level + 1);
//     }

//  }
